Sources/SocksCore/Address.swift

```diff
@@ -167,3 +169,32 @@ extension ResolvedInternetAddress: CustomStringConvertible {
     }
 }
 
+extension Socket {
+    
+    public func remoteAddress() throws -> ResolvedInternetAddress {
+        var length = socklen_t(sizeof(sockaddr_storage))
+        let addr = UnsafeMutablePointer<sockaddr_storage>.init(allocatingCapacity: 1)
+        let addrSockAddr = UnsafeMutablePointer<sockaddr>(addr)
+        let res = getpeername(descriptor, addrSockAddr, &length)
+        guard res > -1 else {
+            addr.deallocateCapacity(1)
+            throw Error(.remoteAddressResolutionFailed)
+        }
+        let clientAddress = ResolvedInternetAddress(raw: addr)
+        return clientAddress
+    }
+    
+    public func localAddress() throws -> ResolvedInternetAddress {
+        var length = socklen_t(sizeof(sockaddr_storage))
+        let addr = UnsafeMutablePointer<sockaddr_storage>.init(allocatingCapacity: 1)
+        let addrSockAddr = UnsafeMutablePointer<sockaddr>(addr)
+        let res = getsockname(descriptor, addrSockAddr, &length)
+        guard res > -1 else {
+            addr.deallocateCapacity(1)
```

Comment:
That's not the same thing. `ResolvedInternetAddress` takes a +1 retained raw pointer, which it just takes over and later deallocates itself. So we need to manually deallocate it here only in the case that `ResolvedInternetAddress` wasn't already created. If I put it in defer, it'd run every time, destroying the memory that `ResolvedInternetAddress` needs. Not too elegant, but I had real hard time to get it all work on both OS X and Linux before I ended up with this solution (there was some stack allocation weirdness on Linux).